File: rust/src/helpers/lcg.rs
pub struct Lcg {
    m: i32,
    a: i32,
    c: i32,
    last_num: i32,
}

impl Lcg {
    pub fn new(seed: i32, m: i32, a: i32, c: i32) -> Self {
        Self {
            m,
            // For the sake of fairness, modulus will always be 2^32 to bypass modulus bias between languages
            a,
            c,
            last_num: seed,
        }
    }

    pub fn next_int(&mut self) -> i32 {
        self.last_num = self.last_num.wrapping_mul(self.a).wrapping_add(self.c); // modulus is treated as 2^32, 32 bit overflow will automatically wrap around
        self.last_num & 0x7FFFFFFF // ignore bit sign bit to ensure non-negative output
    }

    pub fn next_double(&mut self) -> f64 {
        self.next_int() as f64 / 4294967296.0 // 2^32
    }

   
}

File: typescript/helpers/lcg.js
// Linear Congruential Generator (LCG)
export class LCG {
    constructor(seed, m, a, c) {
        this.last_num = seed;
        this.m = m;
        this.a = a;
        this.c = c;
    }
    nextDouble() {
        return this.nextInt() / this.m;
    }
    nextInt() {
        // modulus should be a very slow operation as TypeScript/JavaScript does not have native support for it
        this.last_num = (this.a * this.last_num + this.c) % this.m;
        return this.last_num;
    }
}
